packages/flutter/lib/src/material/radio.dart

```diff
@@ -371,47 +660,102 @@ class Radio<T> extends StatefulWidget {
 
   final _RadioType _radioType;
 
-  bool get _selected => value == groupValue;
-
   @override
   State<Radio<T>> createState() => _RadioState<T>();
 }
 
 class _RadioState<T> extends State<Radio<T>> with TickerProviderStateMixin, ToggleableStateMixin {
   final _RadioPainter _painter = _RadioPainter();
+  _RadioGroupState<T>? _group;
+  _RadioGroupPolicyState<T>? _policy;
+
+  FocusNode? _internalFocusNode;
 
-  void _handleChanged(bool? selected) {
-    if (selected == null) {
-      widget.onChanged!(null);
+  /// The focus node for this radio state.
+  @protected
+  FocusNode get focusNode => widget.focusNode ?? (_internalFocusNode ??= FocusNode());
+
+  bool get _usesRadioGroup => _group != null;
+  bool get _handlesChanges => widget.onChanged != null || _usesRadioGroup;
+
+  T? get _effectiveGroupValue => _group?.groupValue ?? widget.groupValue;
+
+  /// Handle selection status changed.
+  ///
+  /// if `selected` is false, nothing happens.
+  ///
+  /// if `selected` is true, select this radio. i.e. [Radio.onChanged] is called
+  /// with [Radio.value]. This also updates the group value in [RadioGroup] if it
+  /// is in use.
+  ///
+  /// if `selected` is null, unselect this radio. Same as `selected` is true
+  /// except group value is set to null.
+  @protected
+  void handleChanged(bool? selected) {
+    assert(_handlesChanges);
+    if (!(selected ?? true)) {
       return;
     }
-    if (selected) {
-      widget.onChanged!(widget.value);
+    _handleGroupValueChanged(selected ?? false ? widget.value : null);
+  }
+
+  void _handleGroupValueChanged(T? newGroupValue) {
+    if (_usesRadioGroup) {
+      _group!.groupValue = newGroupValue;
+    }
+    if (widget.onChanged != null) {
+      widget.onChanged!(newGroupValue);
     }
   }
 
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _group = context.dependOnInheritedWidgetOfExactType<_RadioGroupStateScope<T>>()?.state;
+    assert(
+      _group == null || widget.groupValue == null,
+      'A RadioGroup can not wrap Radio widget with a non null groupValue. '
```

Comment:
This message was a little confusing, but we can revisit - I know you asked for early feedback.